openapi/openapi/src/main/scala/endpoints/openapi/JsonSchemas.scala

```diff
@@ -35,20 +37,65 @@ trait JsonSchemas extends algebra.JsonSchemas with TuplesSchemas { openapiJsonSc
+    /**
+      * A documented record exists in two flavors, either lazy or strict. All property values in the strict version
+      * are present as abstract `def`s to make it easier in use, regardless of which version you use. The same
+      * principles apply to [[DocumentedCoProd]].
+      */
+    sealed trait DocumentedRecord extends DocumentedJsonSchema {
+      def toStrict: StrictDocumentedRecord
+      def withName(name: String): DocumentedRecord
+    }
+
+    case class StrictDocumentedRecord(
       fields: List[Field],
       additionalProperties: Option[DocumentedJsonSchema] = None,
       name: Option[String] = None,
       example: Option[ujson.Value] = None
-    ) extends DocumentedJsonSchema
+    ) extends DocumentedRecord {
+      val toStrict: StrictDocumentedRecord = this
+      def withName(name: String): DocumentedRecord = copy(name = Some(name))
+    }
+
+    class LazyDocumentedRecord(_record: => StrictDocumentedRecord) extends DocumentedRecord {
+      lazy val toStrict: StrictDocumentedRecord = _record
+      def example: Option[Value] = toStrict.example
+      def withName(name: String): DocumentedRecord = new LazyDocumentedRecord(toStrict.copy(name = Some(name)))
+    }
+
     case class Field(name: String, tpe: DocumentedJsonSchema, isOptional: Boolean, documentation: Option[String])
 
-    case class DocumentedCoProd(
+    /** @see [[DocumentedRecord]]. */
+    sealed trait DocumentedCoProd extends DocumentedJsonSchema {
+      def toStrict: StrictDocumentedCoProd
+      def withName(name: String): DocumentedCoProd
+      def withDiscriminatorName(name: String): DocumentedCoProd
+    }
+
+    case class StrictDocumentedCoProd(
       alternatives: List[(String, DocumentedRecord)],
       name: Option[String] = None,
       discriminatorName: String = defaultDiscriminatorName,
       example: Option[ujson.Value] = None
-    ) extends DocumentedJsonSchema
+    ) extends DocumentedCoProd {
+      val toStrict: StrictDocumentedCoProd = this
+      def withName(name: String): DocumentedCoProd = copy(name = Some(name))
+      def withAlternatives(alternatives: List[(String, DocumentedRecord)]): DocumentedCoProd = copy(alternatives = alternatives)
+      def withDiscriminatorName(name: String): DocumentedCoProd = copy(discriminatorName = name)
+    }
+
+    class LazyDocumentedCoProd(_coProd: => StrictDocumentedCoProd) extends DocumentedCoProd {
+      lazy val toStrict: StrictDocumentedCoProd = _coProd
+      def example: Option[Value] = toStrict.example
+
+      def withName(name: String): DocumentedCoProd = new LazyDocumentedCoProd(toStrict.copy(name = Some(name)))
+      def withDiscriminatorName(name: String): DocumentedCoProd = new LazyDocumentedCoProd(toStrict.copy(discriminatorName = name))
+    }
+
+    class LazySchema(_schema: => DocumentedJsonSchema) extends DocumentedJsonSchema {
+      lazy val value: DocumentedJsonSchema = _schema
+      def example: Option[ujson.Value] = value.example
+    }
```

Comment:
Sorry to be nitpicking, but I would rather not have `example` as a method on `LazySchema`, `LazyDocumentedCoProd` and `LazyDocumentedRecord`. I think it’s clearer to write `.toStrict.example` instead.

Also, do you think we could make `LazySchema` a (sealed) parent trait of both `LazyDocumentedCoProd` and `LazyDocumentedRecord`? This would remove the need for pattern matching on `LazySchema` if we already pattern match on one of those.

Reply:
Sure, NP. Better to take a bit more time than to rush it in. I'll remove `example` from `LazySchema` and check whether I can make it a sealed (parent) trait.

Reply:
So, we actually have an extension on _endpoints_ where we define 

```
def multiLazyJsonSchema[A, C[X] <: Seq[X]](schema: => JsonSchema[A]): JsonSchema[C[A]]
```

I am not sure how we could define it without `LazySchema` since we do not know if it is a tagged or a record. Which is why I brought it back in the first place. Any ideas?

Reply:
May I ask what is your use case for this function?

Also, wouldn’t it also work if `LazySchema` was a parent type of both `LazyDocumentedCoProd` and `LazyDocumentedRecord`?

Reply:
> Sorry to be nitpicking, but I would rather not have `example` as a method on `LazySchema`, `LazyDocumentedCoProd` and `LazyDocumentedRecord`. I think it’s clearer to write `.toStrict.example` instead.

Actually, I’m thinking a bit more about this one, and the `example` member would not be easy to remove because, as you said in another comment, it is required by the `DocumentedJsonSchema` trait.

Also, it was already in `LazySchema` before your PR, so it’s fine to keep it as-is.

Reply:
> May I ask what is your use case for this function?

We have a model that could look roughly like this: `Trip -> Actions -> Shipments -> Actions -> Trip`. Basically the model is a cyclic graph, however various use cases are trees with a different root walking this graph (without cycles). For example, we have trips that have load and unload actions to deliver various shipments on various locations (Trip -> Actions -> Shipments). However from a shipment point of view you could also be part of multiple trips (e.g. from China to some harbor to big town to a small town) (Shipment -> Actions -> Trips).

Technically we could make this a `Tagged` instead of a `JsonSchema` since Actions is a Tagged, so it could probably work out in the end. However it feels less expressive if you have to copy every method for `Tagged` and `Record`, especially if you do not care which it is. Similar to how the non lazy array variant currently works.

> Also, wouldn’t it also work if LazySchema was a parent type of both LazyDocumentedCoProd and LazyDocumentedRecord

Not sure. You'd still need to know whether it is a record or a tagged if you cannot create a `LazySchema` directly. I also don't think there is much harm in keeping it like this. Though I am happy to explore other alternatives.

Reply:
I’m realizing that I should have started with that but, can you show me some code that needs `lazyRecord` to return a `Record` instead of a `JsonSchema`? The idea of `lazyRecord` or `lazyTagged` is that it allows you to refer to a schema that is being defined. However, the specific type of this schema should not matter.

Reply:
```
case class Trip(
  entityInfo: EntityInfo,
  vehicle: Option[Association[Vehicle]] = None,
  actors: Option[Seq[Association[Actor]]] = None,
  actions: Option[Seq[Association[Action]]] = None,
  constraint: Option[Association[Constraint]] = None,
) extends Entity

lazy val tripJsonSchema: Record[Trip] = (
      jsonSchema[EntityInfo] and // jsonSchema is a custom summoner (implicitly)
      lazyOptField[Association[A]]("vehicle", associationJsonSchema(vehicleJsonSchema))) zip
      optField[Seq[Association[Actor]]]("actors")(multiLazyJsonSchema(associationJsonSchema(actorFormat))) zip
      optField[Seq[Association[Action]]]("actions")(multiLazyJsonSchema(associationJsonSchema(actionJsonSchemas)) zip
      lazyOptField[Association[Constraint]]("constraint", associationJsonSchema(constraintJsonSchema))
  ).xmap((Trip.apply _).tupled)(unlift(Trip.unapply))
```

Reply:
I believe you could use `optField` instead of `lazyOptField`, and `arrayJsonSchema` instead of `multiLazyJsonSchema` like so:

~~~ scala
def associationVehicleSchema(tripJsonSchemaRef: JsonSchema[Trip]): JsonSchema[Assocation[Vehicle]] = ...
def associationActorSchema(tripJsonSchemaRef: JsonSchema[Trip]): JsonSchema[Association[Actor]] = ...

val tripJsonSchema: Record[Trip] = {
  def selfRef = lazyRecord(tripJsonSchema, "Trip")

  optField("vehicle")(associationVehicleSchema(selfRef)) zip
  optField("actors")(arrayJsonSchema(associationActorSchema(selfRef)))
}.xmap(…)(…)
~~~

But I agree that it’s not very convenient to write (ie., you can not just let the compiler summon the schema of the optional field `actors`).

What do you think?

Reply:
But these calls will ensure that the schema will be evaluated right? I don't expect that to work with recursive calls. Or am I missing something?

Reply:
Also, thinking about it more. I don't think it makes sense to have lazy constructors explicitly called `lazyRecord` and `lazyTagged` and not return a `Record` or `Tagged` respectively. They won't compose well with how tagged and record are used in other places. I think the encoding does this better than the current solution. Even though we might have to think a bit more on how to deal with other `LazySchema`s.

Reply:
> But these calls will ensure that the schema will be evaluated right? Since `optField` is strict in its argument.

`optField` is strict but it get passed a `lazyRecord`, which does not evaluate the underlying record schema until needed. So, I think this will work.

> I don't think it makes sense to have lazy constructors explicitly called `lazyRecord` and `lazyTagged` and not return a `Record` or `Tagged` respectively.

True. Actually, I don’t remember why I had to introduce this distinction from `lazySchema` to `lazyRecord` and `lazyTagged`. But now I think a single `lazySchema[A](toStrict: => JsonSchema[A]): JsonSchema[A]` should be enough.

Reply:
> optField is strict but it gets passed a lazyRecord, which does not evaluate the underlying record schema until needed. So, I think this will work.

That makes sense. Though I am not sure I like the ergonomics. Because it is not just one schema I now have to pass explicitly, but all of them (every entity points to actions and every action can point back). I also don't really understand what would be against extending the current API. The code doesn't get much more complicated, and the way to use it seems IMO simpler. Maybe you can elaborate a little bit on that? Because it seems you are leaning towards closing this PR?

Reply:
Also I want to point out that I am a little disappointed that after the work I put in this might just be thrown away.

_However_, I also don't think that is a good enough reason to merge it. It should have value to make these changes. So if you don't think it will, I will accept that. So, it's not necessary to feel guilty about just closing it.

Reply:
Hey @bmeesters, I’m sorry this is my fault, it’s only after seeing the PR that I realized this is _maybe_ not the best way to go.

Whatever we do with regards to this PR (closing it or merging it), we should make sure that we find a satisfactory solution for your original issue. In particular:

>  I am not sure I like the ergonomics. Because it is not just one schema I now have to pass explicitly, but all of them (every entity points to actions and every action can point back)

I agree with this. Since `field`, `optField` or `arrayJsonSchema` are type-directed (they take an *implicit* `JsonSchema`), one solution could be to provide a type-level `Lazy` operator:

~~~ scala
type Lazy[A]

def lazySchema[A](toStrict: => JsonSchema[A]): JsonSchema[Lazy[A]]
~~~

And then:

~~~ scala
val tripJsonSchema: Record[Trip] = {
  implicit def selfRef: JsonSchema[Lazy[Trip]] = lazySchema(tripJsonSchema, "Trip")

  optField[Lazy[Trip]]("nextTrip") zip
  optField[List[Lazy[Trip]]]("nextTrips")
}.xmap(…)(…)
~~~

I need to think a bit more about it, to see whether this is a viable solution or not. Do you have any opinions?

Reply:
No problem @julienrf. This is always the benefit of hindsight. We are not our code and I am definitely in favor of finding the best solution. The `Lazy` type approach sounds promising. Though I can't oversee how that would impact for instance all the Json implementations.

I think for now, I'll try to make it work with whatever is currently available (we have some customers waiting on an OpenAPI specification). And its fine by me to let it rest for a bit so you have time to think about it. I would also be willing to work on the Lazy approach. Though I expect I need some pointers along the way.